### PR TITLE
Implement compute shaders support, make fragment shader non-mandatory

### PR DESCRIPTION
**Pull request by VReaperV**
Requires #1124.

Cherry-picked from #1137 to split a large pr up a bit.

Adds some code to allow using compute shaders. Shaders can now have either of vertex, fragment, or compute shaders enabled/disabled.

### CODE REVIEW

File: src/engine/renderer/gl_shader.cpp [hunk 1]
```diff
@@ -770,9 +809,46 @@ std::string     GLShaderManager::BuildGPUShaderText( Str::StringRef mainShaderNa
...
+	std::string line;
+
+	while ( std::getline( shaderTextStream, line, '\n' ) ) {
+		const std::string::size_type position = line.find( "#insert" );
+		if ( position == std::string::npos ) {
+			shaderMain += line + "\n";
+			continue;
+		}
+
+		const std::string::iterator beginIt = std::find_if( line.begin(), line.end(),
```
Comment (slipher) [inline src/engine/renderer/gl_shader.cpp:826]: Could use `find_first_not_of(" \t)`, then you get an unsigned index like `position`.

**Review by slipher — COMMENTED** [review]
It seems like compute shaders are a totally separate workflow from the existing draw shaders, which would ideally be a separate class. Like instead of just GLShader you have one class with fragment shader, vertex shader, and vertex attribs; and another with just compute shader.

**Comment by VReaperV** [comment]
> It seems like compute shaders are a totally separate workflow from the existing draw shaders, which would ideally be a separate class. Like instead of just GLShader you have one class with fragment shader, vertex shader, and vertex attribs; and another with just compute shader.

I considered that, the reason I made it part of GLShader is that they can still make use of the way uniforms are set, vertex/fragment includes, same paths for loading/compiling shaders including shader binaries, as well as GLHeaders. Maybe both classes should be derived from GLShader which would just provide the afornentioned functionality.

**Comment by slipher** [comment]
A really easy way to make things a bit clearer could be to have two constructors for GLShader, one for each use case. Instead of having a constructor that allows all 3 shader types at once.

**Comment by VReaperV** [comment]
So something like `GLShader( const std::string &name, const std::string &mainShaderName, uint32_t vertexAttribsRequired, GLShaderManager *manager, const bool hasVertexShader, const bool hasFragmentShader)` and `GLShader( const std::string &name, const std::string &mainShaderName, uint32_t vertexAttribsRequired, GLShaderManager *manager, const bool hasComputeShader`?

**Comment by slipher** [comment]
I guess there's nothing that distinguishes the constructors very well so maybe it would be easier to go ahead and define the subclasses even if all the functionality is in the base class. Like

```
class GLShaderBase {
    // everything that's in GLShader right now
};
class GLShader : public GLShaderBase { //draw shader
   // constructor does _hasFragment = true, _haveVertex = true?
};
class ComputeShader : public GLShaderBase {
    // constructor does _hasCompute = true
};

**Comment by VReaperV** [comment]
Yeah, I ran into some issues with GLShader constructors earlier too, since the wrong one was being selected with some types being converted.

File: src/engine/renderer/gl_shader.cpp (hunk 1 above)
Comment (VReaperV) [reply src/engine/renderer/gl_shader.cpp]: Done now.

**Comment by VReaperV** [comment]
> I guess there's nothing that distinguishes the constructors very well so maybe it would be easier to go ahead and define the subclasses even if all the functionality is in the base class. Like
> 
> ```
> class GLShaderBase {
>     // everything that's in GLShader right now
> };
> class GLShader : public GLShaderBase { //draw shader
>    // constructor does _hasFragment = true, _haveVertex = true?
> };
> class ComputeShader : public GLShaderBase {
>     // constructor does _hasCompute = true
> };
> ```

This should be evaluated later because #1105 and some further PRs add more GLShader constructors.

**Review by illwieckz — APPROVED** [review]
LGTM.

**Comment by illwieckz** [comment]
Don't forget to squash the `fix` commit. ;-)

**Comment by VReaperV** [comment]
Yep, done.

**Pull request merged**